### PR TITLE
Add script create-assigned-issue

### PR DESCRIPTION
**Pull request by jaw6**
This is intended as a companion to [add-collaborators](https://github.com/jaw6/training-utils/blob/a8042c751b41f4d9dab86981da573fe2c776fd2e/add-collaborators.rb).

The idea is that this might be used, during class, to create & assign an issue per student, based on a target/template issue.

So: point `create-assigned-issue` at an issue. This issue becomes the "template". The script will create new issues, one for each repo collaborator, based on the template. Any done todo items will be undone, and any references within the template to its assigned user will be rewritten for the newly assigned user.

I'm thinking that the template would be an issue we create and assign to @githubstudent. So, any references inside that issue to `@githubstudent` (or `githubstudent`) will be re-written. Here's an example of how this might work:
### Target/template issue:

**Title:** Create bio for githubstudent
**Body:**

```
* [x] Create new branch called githubstudent
* [x] Add a new file githubstudent.md
* [x] Create a pull request, referencing this issue by number
* [x] At-mention @githubteacher
```

**Assigned:** @githubstudent
### Turns into:

**Title:** Create bio for jaw6
**Body:**

```
* [ ] Create new branch called jaw6
* [ ] Add a new file jaw6.md
* [ ] Create a pull request, referencing this issue by number
* [ ] At-mention @githubteacher
```

**Assigned:** @jaw6

---

Thoughts? Ideas? Questions? Feedback?

### CODE REVIEW

**Comment by matthewmccullough** [comment]
@jaw6 I love this type of utility for teaching.  This is super awesome work. :heart: :sparkles:

**Comment by matthewmccullough** [comment]
CC @PeterBell just as an FYI for your efforts in creating bots and automation within Wheelhouse for setting up a "clean slate" for each student in class.

**Comment by PeterBell** [comment]
Thanks @matthewmccullough - shared!

**Comment by brntbeer** [comment]
I love this! One worry is how spammy it'll get though =( Any thoughts around that?

File: create-assigned-issue.rb [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+$stderr.sync = true
+require 'octokit'
+require 'optparse'
+
+file      = __FILE__
+issue_url = ARGV.first
+token     = ENV['TOKEN']
+
+abort banner unless issue_url || issue_url == '-h'
```
Comment (brntbeer) [inline create-assigned-issue.rb]: This is real slick. I like this a lot.

File: create-assigned-issue.rb [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+Usage: create-assigned-issue.rb <issue URL>
+  * Copies issue as template
+  * Replaces any references to the assigned user (eg, at-mentions, etc.)
+  * Also "undoes" any checkboxes in issue template (eg, [x] -> [ ])
+  * For each collaborator of repo, creates a new issue, assigned to collaborator
+BANNER
+
+$stderr.sync = true
+require 'octokit'
+require 'optparse'
```
Comment (brntbeer) [inline create-assigned-issue.rb:12]: Would it make more sense for this to be at the very top of the file? May just be me but that's where i go for finding libraries a project uses

Comment (jaw6) [inline create-assigned-issue.rb:12]: Yeah, I kinda wanted to do something like [this](https://gist.github.com/rtomayko/1190547) but it wound up feeling like overkill. But the inspiration was: first explain what this does (but in a way that's re-usable in the script `-h` output), then do some stuff. I dunno. I feel like _everything_ wants to be "at the top" but there's only one "top". 

If nothing else, once we pick a "winner" maybe I should update `add-collaborators` to follow the same style. Like, I hear people like consistent style or something?

File: create-assigned-issue.rb (hunk 1 above)
Comment (jaw6) [inline create-assigned-issue.rb]: Although! Turned out it didn't actually like really work? Should work now, though.

File: create-assigned-issue.rb (hunk 2 above)
Comment (brntbeer) [inline create-assigned-issue.rb:12]: hahaha consistency.

Yeah, that makes sense and it does read real well for that. Okay good motivation. Thanks!

**Comment by jaw6** [comment]
> One worry is how spammy it'll get though =( Any thoughts around that?

So, I think we're already seeing a little of that in the GH4Devs curriculum -- add students as collaborators and pretty much spam them with a ton of notifications straight away. For better or worse, in class, we then use this as an opportunity to talk about Notifications & Notification settings -- personally, I think it works. 

I guess, in the context of classes as we're currently approaching them, this script doesn't strike me as _adding_ much extra spam -- they're already collaborators at this point. 

This is where @crichID and I were talking about this a few weeks back: I believe the curriculum right now has us asking them to create an issue, assign to themselves, then edit & PR targeting their issue. In the Intro class, especially, it feels like a lot. And, I think, in the 4Devs class, it feels like there's some extraneous hat-switching we're asking of the students. This might just be a way to boost a class that's flagging a bit behind the expected pace.

Also, as far as my intent goes, I made it this way so that it's optional -- I'm not suggesting we build class around the script at all, but if it works to accelerate or streamline our process (like I think it will with Intro) we _can_.

**Comment by matthewmccullough** [comment]
One thing to consider is fan-out of @mentions on Issues.  Both from my classroom experience, as well as our ever-more successful online classes, I'm convinced that an Issue-per-student is the right path and less noisy than Issue-per-class.  The robot (CC @PeterBell, Wheelhouse) will help give each student a "sandboxed" and more managed interactive experience.  In fact, I really like this taking a step further and making a repo in each user's account per class so that they get to "keep it" after class, and that it doesn't noise up or get throw away like we do when we reset `githubschool` repos.

**Comment by PeterBell** [comment]
@matthewmccullough definitely want to chat with @crichID about this. I was teaching a class the other day and while I'm reluctant to lose the social benefits of a class repo, I consistently find that when you ask everyone to create a PR 80% do. Then when you ask everyone to comment on the PR below theirs, maybe half the people do - and some of them pick the wrong PR, so some people have no comments, some have two. Some people close other peoples PRs, other people create PR's for someone elses branch - in short it becomes a fairly confusing learning experience for the students.

I always liked the idea of class repo's but I'm starting to think on balance that generating individual repos and just having bots interact with everyone gives a more consistent learning experience. Will chat with @crichID and see if we can maybe run some experiments around this.

**Comment by PeterBell** [comment]
And generating individual repo's is what we do for all self serve, so we already have that capability.

**Comment by jaw6** [comment]
Maybe it's just the way I've been thinking about these things alone but I
think if I went to a GitHub class and spent the whole time working solo I
might be disappointed in that outcome. A big part of the GitHub branding is
working together, social coding etc.

I think I have been assuming that our goals were to use scripts and robots
to enhance and emulate an active group collaboration environment more than
to isolate and separate students. But maybe I have that wrong.

**Comment by PeterBell** [comment]
@jaw6 here's the question (and I don't have the answer). I agree you have to be working and interacting with other people. Do those people have to be humans?

**Comment by crichID** [comment]
> This is where @crichID and I were talking about this a few weeks back: I believe the curriculum right now has us asking them to create an issue, assign to themselves, then edit & PR targeting their issue.

To add some additional context here. This conversation was originally about how we could pare down the Intro to GitHub Course so it would feel less frantic for the learners. The idea was to use the issue creation template and remove the requirement for them to create an issue since it is a nominal skill. @jaw6 - thank you for taking this on and doing a :sparkles: job!

> For better or worse, in class, we then use this as an opportunity to talk about Notifications & Notification settings -- personally, I think it works.

I agree. It is an unfortunate by-product of the process but leads to a memorable learning experience!

> I always liked the idea of class repo's but I'm starting to think on balance that generating individual repos and just having bots interact with everyone gives a more consistent learning experience.

@PeterBell I know we have spoken about this several times in the past, but I still believe the social aspect of the learning experience trumps the need for perfection in class exercises. We are social creatures and are far more likely to engage when we know our peers have visibility into our actions. I think creating separate repos will actually compound the issues you mention re: participation instead of reduce them.

> Do those people have to be humans?

Yes - humans are not emotionally compelled to respond to robots (yet)!

> A big part of the GitHub branding is working together, social coding etc.

@jaw6 :dart: people are not paying to learn how to click buttons. They are paying to learn a better way of working. Does Coke sell a sugary beverage or happiness?

**Comment by matthewmccullough** [comment]
Let me know if there's any historical experience or sentiments that I can share here to help us move this PR forward to a place that most of @github/training-teachers are happy with. :smile:

**Comment by crichID** [comment]
> I always liked the idea of class repo's but I'm starting to think on balance that generating individual repos and just having bots interact with everyone gives a more consistent learning experience.

@PeterBell Unfortunately, we are having an even harder time getting them to log in to certify. During yesterday's GH for Dev class online, I could only get ~35 students of the 76 in class to sign in to Certify - despite begging and encouraging them to do so for 5 minutes. On the other hand, 52 signed up as collaborators using the issue (which we were forced to do because Certify didn't add them as collaborators to the repo). 

I think the idea of bots is appealing, but I think we also need to consider that learning is largely a social experience, as is GitHub. People may expect to interact with bots in self-serve training, but that is not why they are paying for facilitator led instruction.

@jaw6 what do you need from me in order to finish up this script? I am teaching Intro to GitHub on Wednesday and I would like to use it in that class!

**Comment by patrickmckenna** [comment]
I think there's quite a bit in this thread that we could and should keep discussing. That said, I see no reason to hold off shipping this particular PR, @jaw6. It always helps to have another option, and we're not requiring that anyone use it.

**Comment by jaw6** [comment]
> what do you need from me in order to finish up this script?

Right now, this is at "works on my machine" -- I'd :heart: someone (anyone) to confirm that it works for them and then I think (I _think_) we're good to :ship:

**Comment by PeterBell** [comment]
> we are having an even harder time getting them to log in to certify

Hi @crichid I think that's the easiest one to fix. We need to rework the flow so students get a Wheelhouse login to go to. When they login on the first page it'll show the GTW info. Getting Wheelhouse to work for people that don't log into it is like trying to get people who don't log into GTW to learn from the class - impossible. We'll need to add another field to the form, and as per your previous discussion break out the forms for the different event types to make them less confusing.

FYI, the main problem we had with the last class was there was a leading space in the token which meant the bots couldn't log in to add people as collaborators. But in good news, we had logging, saw it almost immediately and were able to resolve it.

First couple of classes were bumpy with issues with pusher and problems which we've identified and fixed with the labs. Next step when you're back in the office Friday or next week, lets get a gameplan to get you comfortable that things are now working (which they now appear to be). Once we've done that, we can then rework the workflow for an upcoming class to make sure people log in via Wheelhouse to get the GTW URL. Definitely don't want paying customers to feel like they're being experimented on, so let's discuss what we need to do for you to be 100% comfortable with the instructor led experience - perhaps pairing on a class of people who are not GitHub enterprise clients just to prove we're solid.

**Comment by matthewmccullough** [comment]
> perhaps pairing on a class of people who are not GitHub enterprise clients just to prove we're solid.

:+1: I like this pairing idea to help get us that "Wheelhouse" insider view and confidence. :sparkles:

**Comment by patrickmckenna** [comment]
> Right now, this is at "works on my machine" -- I'd  someone (anyone) to confirm that it works for them and then I think (I think) we're good to

@jaw6 :+1: Just tested it on a personal repo of mine, with githubteacher and githubstudent added as collaborators, and got two new Issues w/ all the desired properties (new @mentions, assignee, &c.).

**Comment by jaw6** [comment]
I've gone back-and-forth on whether this was ready to merge & close a couple of times now, reading & re-reading everyone's comments, and waiting to see if the conversation was still on-going. At this point, I think it's a little over-due: I'm going to merge now and close down the PR, but feel free to open an issue or revert or whatever if folks _are_ interested in continuing the conversation.

**Comment by jaw6** [comment]
Er, actually, no. I forgot that I'm apparently no longer a collaborator of this repo, post move/re-org. Can someone maybe collaboratize me? (@brntbeer?)

**Comment by brntbeer** [comment]
added @github/training-teachers ! merge away sir

**Pull request merged**